src/Ocelot.Provider.Consul/ConsulProviderFactory.cs

```diff
@@ -17,16 +18,20 @@ public static class ConsulProviderFactory
 
     public static ServiceDiscoveryFinderDelegate Get { get; } = CreateProvider;
 
+    private static ConsulRegistryConfiguration configuration;
+    private static ConsulRegistryConfiguration ConfigurationGetter() => configuration;
+    public static Func<ConsulRegistryConfiguration> GetConfiguration { get; } = ConfigurationGetter;
+
     private static IServiceDiscoveryProvider CreateProvider(IServiceProvider provider,
         ServiceProviderConfiguration config, DownstreamRoute route)
     {
         var factory = provider.GetService<IOcelotLoggerFactory>();
         var consulFactory = provider.GetService<IConsulClientFactory>();
 
-        var consulRegistryConfiguration = new ConsulRegistryConfiguration(
-            config.Scheme, config.Host, config.Port, route.ServiceName, config.Token);
+        configuration = new ConsulRegistryConfiguration(config.Scheme, config.Host, config.Port, route.ServiceName, config.Token);
+        var serviceBuilder = provider.GetService<IConsulServiceBuilder>();
 
-        var consulProvider = new Consul(consulRegistryConfiguration, factory, consulFactory);
+        var consulProvider = new Consul(configuration, factory, consulFactory, serviceBuilder);
 
         if (PollConsul.Equals(config.Type, StringComparison.OrdinalIgnoreCase))
```

Comment:
The problem then... What is a Consul Service Provider? How should we handle cases like the Long Polling? I am preparing a PR implementing the consul long polling capabilities. Where should I place that?  I have implemented it, as it would be a standard functionality of the consul service provider, since I think that the Matroshka logic we have implemented here is not great, having an IServiceDiscoveryProvider embedded in another IServiceDiscoveryProvider...

Reply:
Tom proposed the Matroshka logic. 🙈
I concur with 99% of the sentiment that the `ConsulProviderFactory` requires refactoring and should be transformed into a genuine DI-interface with a specific implementation class. The current `static` class implementation is indeed unattractive.

> What is a Consul Service Provider? How should we handle cases like the Long Polling?

At the moment, I'm not clear on this issue.
We have two classes of providers: `Consul` and `PollConsul`, with the following responsibilities:
- `Consul` encapsulates an `IConsulClient` object for integration with a Consul server instance. This provider executes HTTP requests to access the Consul registry.
- `PollConsul` encapsulates an `IServiceDiscoveryProvider` object, which is `Consul` at runtime. Therefore, it is tasked with service renewal through a triggering mechanism.

Will the `Consul` and `PollConsul` classes be retained in your envisioned future design?
Once more, the aim of this PR is to address the issue of multiple Consul setups through a virtual mechanism in `Consul` and `DefaultConsulServiceBuilder`, which I believe is an acceptable bug fix.
I also appreciate your desire to further develop the design of the Consul provider.

The Matroshka logic is slated for removal in the upcoming refactoring phases. 😋
Apologies if the changes have caused merge conflicts in your current feature branches, which I am unable to see. 😛